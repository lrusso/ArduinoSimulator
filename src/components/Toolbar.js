import React from "react"
import IconNew from "../assets/IconNew"
import IconOpen from "../assets/IconOpen"
import IconSave from "../assets/IconSave"
import IconUndo from "../assets/IconUndo"
import IconRedo from "../assets/IconRedo"
import IconSearch from "../assets/IconSearch"
import IconStart from "../assets/IconStart"
import IconStop from "../assets/IconStop"
import ToolbarItem from "./ToolbarItem"
import ToolbarBoard from "./ToolbarBoard"
import ToolbarFilename from "./ToolbarFilename"
import ToolbarSeparator from "./ToolbarSeparator"
import Loading from "./Loading"
import ConfirmBox from "./ConfirmBox"
import { useSimulatorContext } from "../contexts/SimulatorContext"
import {
  editorNew,
  editorSetValue,
  editorSave,
  editorUndo,
  editorRedo,
  editorSearch,
  editorIsDirty,
  editorEnable,
  editorDisable,
  editorFocus,
} from "../utils/editor"
import { startSimulator, stopSimulator } from "../utils/interpreter"
import {
  BOARD_MEGA1280,
  BOARD_MEGA2560,
  BOARD_NANO,
  BOARD_UNO,
  t,
} from "../utils/service"

const Toolbar = () => {
  const [showConfirmMessage, setShowConfirmMessage] = React.useState(null)
  const [showLoading, setShowLoading] = React.useState(false)
  const {
    simulatorRunning,
    setSimulatorRunning,
    filename,
    setFilename,
    boardType,
    setBoardType,
    setDigitalPin0,
    setDigitalPin1,
    setDigitalPin2,
    setDigitalPin3,
    setDigitalPin4,
    setDigitalPin5,
    setDigitalPin6,
    setDigitalPin7,
    setDigitalPin8,
    setDigitalPin9,
    setDigitalPin10,
    setDigitalPin11,
    setDigitalPin12,
    setDigitalPin13,
    setDigitalPin14,
    setDigitalPin15,
    setDigitalPin16,
    setDigitalPin17,
    setDigitalPin18,
    setDigitalPin19,
    setDigitalPin20,
    setDigitalPin21,
    setDigitalPin22,
    setDigitalPin23,
    setDigitalPin24,
    setDigitalPin25,
    setDigitalPin26,
    setDigitalPin27,
    setDigitalPin28,
    setDigitalPin29,
    setDigitalPin30,
    setDigitalPin31,
    setDigitalPin32,
    setDigitalPin33,
    setDigitalPin34,
    setDigitalPin35,
    setDigitalPin36,
    setDigitalPin37,
    setDigitalPin38,
    setDigitalPin39,
    setDigitalPin40,
    setDigitalPin41,
    setDigitalPin42,
    setDigitalPin43,
    setDigitalPin44,
    setDigitalPin45,
    setDigitalPin46,
    setDigitalPin47,
    setDigitalPin48,
    setDigitalPin49,
    setDigitalPin50,
    setDigitalPin51,
    setDigitalPin52,
    setDigitalPin53,
    setAnalogPin0,
    setAnalogPin1,
    setAnalogPin2,
    setAnalogPin3,
    setAnalogPin4,
    setAnalogPin5,
    setAnalogPin6,
    setAnalogPin7,
    setAnalogPin8,
    setAnalogPin9,
    setAnalogPin10,
    setAnalogPin11,
    setAnalogPin12,
    setAnalogPin13,
    setAnalogPin14,
    setOutputData,
  } = useSimulatorContext()
  const refUploader = React.useRef(null)
  const NEW_FILE = 1
  const OPEN_FILE = 2

  const newFile = () => {
    if (editorIsDirty()) {
      editorDisable()
      setShowConfirmMessage(NEW_FILE)
    } else {
      cleanEditor()
    }
  }

  const openFile = () => {
    if (editorIsDirty()) {
      editorDisable()
      setShowConfirmMessage(OPEN_FILE)
    } else {
      uploadFile()
    }
  }

  const openFileConfirmed = async (event) => {
    const selectedFileForUploading = event.target.files[0]
    const documentName = selectedFileForUploading.name
    setFilename(documentName)

    const filereader = new FileReader()
    filereader.onload = function () {
      editorSetValue(this.result)
      editorFocus()
      refUploader.current.value = null
    }
    filereader.readAsText(selectedFileForUploading)
  }

  const saveFile = () => {
    editorSave(filename)
  }

  const cleanEditor = () => {
    stopSimulator()
    setSimulatorRunning(false)
    setDigitalPin0(false)
    setDigitalPin1(false)
    setDigitalPin2(false)
    setDigitalPin3(false)
    setDigitalPin4(false)
    setDigitalPin5(false)
    setDigitalPin6(false)
    setDigitalPin7(false)
    setDigitalPin8(false)
    setDigitalPin9(false)
    setDigitalPin10(false)
    setDigitalPin11(false)
    setDigitalPin12(false)
    setDigitalPin13(false)
    setDigitalPin14(false)
    setDigitalPin15(false)
    setDigitalPin16(false)
    setDigitalPin17(false)
    setDigitalPin18(false)
    setDigitalPin19(false)
    setDigitalPin20(false)
    setDigitalPin21(false)
    setDigitalPin22(false)
    setDigitalPin23(false)
    setDigitalPin24(false)
    setDigitalPin25(false)
    setDigitalPin26(false)
    setDigitalPin27(false)
    setDigitalPin28(false)
    setDigitalPin29(false)
    setDigitalPin30(false)
    setDigitalPin31(false)
    setDigitalPin32(false)
    setDigitalPin33(false)
    setDigitalPin34(false)
    setDigitalPin35(false)
    setDigitalPin36(false)
    setDigitalPin37(false)
    setDigitalPin38(false)
    setDigitalPin39(false)
    setDigitalPin40(false)
    setDigitalPin41(false)
    setDigitalPin42(false)
    setDigitalPin43(false)
    setDigitalPin44(false)
    setDigitalPin45(false)
    setDigitalPin46(false)
    setDigitalPin47(false)
    setDigitalPin48(false)
    setDigitalPin49(false)
    setDigitalPin50(false)
    setDigitalPin51(false)
    setDigitalPin52(false)
    setDigitalPin53(false)
    setAnalogPin0(0)
    setAnalogPin1(0)
    setAnalogPin2(0)
    setAnalogPin3(0)
    setAnalogPin4(0)
    setAnalogPin5(0)
    setAnalogPin6(0)
    setAnalogPin7(0)
    setAnalogPin8(0)
    setAnalogPin9(0)
    setAnalogPin10(0)
    setAnalogPin11(0)
    setAnalogPin12(0)
    setAnalogPin13(0)
    setAnalogPin14(0)
    setFilename(null)
    editorNew()
    setShowConfirmMessage(null)
  }

  const uploadFile = () => {
    stopSimulator()
    setDigitalPin0(false)
    setDigitalPin1(false)
    setDigitalPin2(false)
    setDigitalPin3(false)
    setDigitalPin4(false)
    setDigitalPin5(false)
    setDigitalPin6(false)
    setDigitalPin7(false)
    setDigitalPin8(false)
    setDigitalPin9(false)
    setDigitalPin10(false)
    setDigitalPin11(false)
    setDigitalPin12(false)
    setDigitalPin13(false)
    setDigitalPin14(false)
    setDigitalPin15(false)
    setDigitalPin16(false)
    setDigitalPin17(false)
    setDigitalPin18(false)
    setDigitalPin19(false)
    setDigitalPin20(false)
    setDigitalPin21(false)
    setDigitalPin22(false)
    setDigitalPin23(false)
    setDigitalPin24(false)
    setDigitalPin25(false)
    setDigitalPin26(false)
    setDigitalPin27(false)
    setDigitalPin28(false)
    setDigitalPin29(false)
    setDigitalPin30(false)
    setDigitalPin31(false)
    setDigitalPin32(false)
    setDigitalPin33(false)
    setDigitalPin34(false)
    setDigitalPin35(false)
    setDigitalPin36(false)
    setDigitalPin37(false)
    setDigitalPin38(false)
    setDigitalPin39(false)
    setDigitalPin40(false)
    setDigitalPin41(false)
    setDigitalPin42(false)
    setDigitalPin43(false)
    setDigitalPin44(false)
    setDigitalPin45(false)
    setDigitalPin46(false)
    setDigitalPin47(false)
    setDigitalPin48(false)
    setDigitalPin49(false)
    setDigitalPin50(false)
    setDigitalPin51(false)
    setDigitalPin52(false)
    setDigitalPin53(false)
    setAnalogPin0(0)
    setAnalogPin1(0)
    setAnalogPin2(0)
    setAnalogPin3(0)
    setAnalogPin4(0)
    setAnalogPin5(0)
    setAnalogPin6(0)
    setAnalogPin7(0)
    setAnalogPin8(0)
    setAnalogPin9(0)
    setAnalogPin10(0)
    setAnalogPin11(0)
    setAnalogPin12(0)
    setAnalogPin13(0)
    setAnalogPin14(0)
    setSimulatorRunning(false)
    refUploader.current.click()
    setShowConfirmMessage(null)
  }

  const hideConfirmMessage = () => {
    editorEnable()
    setShowConfirmMessage(null)
    editorFocus()
  }

  const switchBoard = () => {
    switch (boardType) {
      case null:
        setBoardType(BOARD_MEGA1280)
        break

      case BOARD_UNO:
        setBoardType(BOARD_MEGA1280)
        break

      case BOARD_MEGA1280:
        setBoardType(BOARD_MEGA2560)
        break

      case BOARD_MEGA2560:
        setBoardType(BOARD_NANO)
        break

      case BOARD_NANO:
        setBoardType(BOARD_UNO)
        break

      default:
        break
    }
  }

  const startSketch = () => {
    setShowLoading(true)
    editorDisable()
    startSimulator(
      setShowLoading,
      setSimulatorRunning,
      setDigitalPin0,
      setDigitalPin1,
      setDigitalPin2,
      setDigitalPin3,
      setDigitalPin4,
      setDigitalPin5,
      setDigitalPin6,
      setDigitalPin7,
      setDigitalPin8,
      setDigitalPin9,
      setDigitalPin10,
      setDigitalPin11,
      setDigitalPin12,
      setDigitalPin13,
      setDigitalPin14,
      setDigitalPin15,
      setDigitalPin16,
      setDigitalPin17,
      setDigitalPin18,
      setDigitalPin19,
      setDigitalPin20,
      setDigitalPin21,
      setDigitalPin22,
      setDigitalPin23,
      setDigitalPin24,
      setDigitalPin25,
      setDigitalPin26,
      setDigitalPin27,
      setDigitalPin28,
      setDigitalPin29,
      setDigitalPin30,
      setDigitalPin31,
      setDigitalPin32,
      setDigitalPin33,
      setDigitalPin34,
      setDigitalPin35,
      setDigitalPin36,
      setDigitalPin37,
      setDigitalPin38,
      setDigitalPin39,
      setDigitalPin40,
      setDigitalPin41,
      setDigitalPin42,
      setDigitalPin43,
      setDigitalPin44,
      setDigitalPin45,
      setDigitalPin46,
      setDigitalPin47,
      setDigitalPin48,
      setDigitalPin49,
      setDigitalPin50,
      setDigitalPin51,
      setDigitalPin52,
      setDigitalPin53,
      setAnalogPin0,
      setAnalogPin1,
      setAnalogPin2,
      setAnalogPin3,
      setAnalogPin4,
      setAnalogPin5,
      setAnalogPin6,
      setAnalogPin7,
      setAnalogPin8,
      setAnalogPin9,
      setAnalogPin10,
      setAnalogPin11,
      setAnalogPin12,
      setAnalogPin13,
      setAnalogPin14,
      setOutputData
    )
  }

  const stopSketch = () => {
    stopSimulator()
    setDigitalPin0(false)
    setDigitalPin1(false)
    setDigitalPin2(false)
    setDigitalPin3(false)
    setDigitalPin4(false)
    setDigitalPin5(false)
    setDigitalPin6(false)
    setDigitalPin7(false)
    setDigitalPin8(false)
    setDigitalPin9(false)
    setDigitalPin10(false)
    setDigitalPin11(false)
    setDigitalPin12(false)
    setDigitalPin13(false)
    setDigitalPin14(false)
    setDigitalPin15(false)
    setDigitalPin16(false)
    setDigitalPin17(false)
    setDigitalPin18(false)
    setDigitalPin19(false)
    setDigitalPin20(false)
    setDigitalPin21(false)
    setDigitalPin22(false)
    setDigitalPin23(false)
    setDigitalPin24(false)
    setDigitalPin25(false)
    setDigitalPin26(false)
    setDigitalPin27(false)
    setDigitalPin28(false)
    setDigitalPin29(false)
    setDigitalPin30(false)
    setDigitalPin31(false)
    setDigitalPin32(false)
    setDigitalPin33(false)
    setDigitalPin34(false)
    setDigitalPin35(false)
    setDigitalPin36(false)
    setDigitalPin37(false)
    setDigitalPin38(false)
    setDigitalPin39(false)
    setDigitalPin40(false)
    setDigitalPin41(false)
    setDigitalPin42(false)
    setDigitalPin43(false)
    setDigitalPin44(false)
    setDigitalPin45(false)
    setDigitalPin46(false)
    setDigitalPin47(false)
    setDigitalPin48(false)
    setDigitalPin49(false)
    setDigitalPin50(false)
    setDigitalPin51(false)
    setDigitalPin52(false)
    setDigitalPin53(false)
    setAnalogPin0(0)
    setAnalogPin1(0)
    setAnalogPin2(0)
    setAnalogPin3(0)
    setAnalogPin4(0)
    setAnalogPin5(0)
    setAnalogPin6(0)
    setAnalogPin7(0)
    setAnalogPin8(0)
    setAnalogPin9(0)
    setAnalogPin10(0)
    setAnalogPin11(0)
    setAnalogPin12(0)
    setAnalogPin13(0)
    setAnalogPin14(0)
    setSimulatorRunning(false)
  }

  React.useEffect(() => {
    const styleNode = document.createElement("style")
    const styleText = `
      .arduinosimulator_menu_item:hover{background-color:#E3E3E3 !important;border:thin solid #D3D3D3 !important;cursor:pointer !important}
      @media (pointer: coarse) { .arduinosimulator_menu_item:hover{background-color:#F2F2F2 !important;border:thin solid #F2F2F2 !important}
    `
    const styleTextNode = document.createTextNode(styleText)
    styleNode.appendChild(styleTextNode)
    document.getElementsByTagName("head")[0].appendChild(styleNode)
  }, [])

  return (
    <>
      <div style={styles.container}>
        <div style={styles.menu_scroll}>
          <div style={styles.menu_wrapper}>
            <ToolbarItem onClick={newFile}>
              <IconNew width={16} height={16} />
            </ToolbarItem>
            <ToolbarItem onClick={openFile}>
              <IconOpen width={16} height={16} />
            </ToolbarItem>
            <ToolbarItem onClick={saveFile}>
              <IconSave width={16} height={16} />
            </ToolbarItem>
            <ToolbarSeparator />
            <ToolbarItem onClick={editorUndo}>
              <IconUndo width={16} height={16} />
            </ToolbarItem>
            <ToolbarItem onClick={editorRedo}>
              <IconRedo width={16} height={16} />
            </ToolbarItem>
            <ToolbarSeparator />
            <ToolbarItem onClick={editorSearch}>
              <IconSearch width={24} height={24} />
            </ToolbarItem>
            <ToolbarSeparator />
            <ToolbarItem
              onClick={simulatorRunning ? undefined : switchBoard}
              disabled={simulatorRunning}
            >
              <ToolbarBoard
                boardType={boardType}
                simulatorRunning={simulatorRunning}
              />
            </ToolbarItem>
            <ToolbarSeparator />
            {simulatorRunning && (
              <ToolbarItem onClick={stopSketch}>
                <IconStop width={17} height={17} />
              </ToolbarItem>
            )}
            {!simulatorRunning && (
              <ToolbarItem onClick={startSketch}>
                <IconStart width={17} height={17} />
              </ToolbarItem>
            )}
            <ToolbarSeparator />
            <ToolbarFilename />
          </div>
        </div>
      </div>
      <input
        ref={refUploader}
        type="file"
        style={styles.uploader}
        onChange={openFileConfirmed}
      />
      {showLoading && <Loading />}
      {showConfirmMessage && (
        <ConfirmBox
          title={t("LOSECHANGES_TITLE")}
          message={t("LOSECHANGES_MESSAGE")}
          accept={t("LOSECHANGES_YES")}
          acceptCallback={showConfirmMessage === NEW_FILE ? cleanEditor : uploadFile}
          cancel={t("LOSECHANGES_NO")}
          cancelCallback={hideConfirmMessage}
        />
      )}
    </>
  )
}

const styles = {
  container: {
    height: "40px",
    borderBottom: "thin solid #D3D3D3",
    overflowY: "hidden",
  },
  menu_scroll: {
    backgroundColor: "#F2F2F2",
    left: 0,
    right: 0,
    paddingTop: 0,
    paddingBottom: 0,
    height: "80px",
    overflowX: "scroll",
    overflowY: "hidden",
    outline: "none",
    textAlign: "center",
    fontFamily: "Arial",
    fontSize: "13px",
  },
  menu_wrapper: {
    float: "left",
    width: "550px",
  },
  uploader: {
    display: "none",
  },
}

export default Toolbar
