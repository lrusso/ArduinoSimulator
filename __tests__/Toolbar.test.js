import React from "react"
import renderer from "react-test-renderer"
import Toolbar from "../src/components/Toolbar"

describe("Toolbar", () => {
  test("Rendering 1", () => {
    const tree = renderer.create(<Toolbar />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
