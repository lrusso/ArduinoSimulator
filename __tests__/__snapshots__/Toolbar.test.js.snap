// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar Rendering 1 1`] = `
[
  <div
    style={
      {
        "borderBottom": "thin solid #D3D3D3",
        "height": "40px",
        "overflowY": "hidden",
      }
    }
  >
    <div
      style={
        {
          "backgroundColor": "#F2F2F2",
          "fontFamily": "Arial",
          "fontSize": "13px",
          "height": "80px",
          "left": 0,
          "outline": "none",
          "overflowX": "scroll",
          "overflowY": "hidden",
          "paddingBottom": 0,
          "paddingTop": 0,
          "right": 0,
          "textAlign": "center",
        }
      }
    >
      <div
        style={
          {
            "float": "left",
            "width": "550px",
          }
        }
      >
        <div
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "float": "left",
              "height": "40px",
              "margin": 0,
            }
          }
        >
          <div
            className="arduinosimulator_menu_item"
            onClick={[Function]}
            style={
              {
                "MozUserSelect": "none",
                "WebkitUserSelect": "none",
                "backgroundColor": "#F2F2F2",
                "border": "thin solid #F2F2F2",
                "cursor": "default",
                "display": "block",
                "fontFamily": "Arial",
                "fontSize": "15px",
                "height": "28px",
                "lineHeight": "32px",
                "marginLeft": "3px",
                "paddingBottom": "1px",
                "paddingTop": "3px",
                "userSelect": "none",
                "width": "28px",
              }
            }
          >
            <svg
              height={16}
              viewBox="0 0 1000 1000"
              width={16}
            >
              <path
                d="M903 761l-142 142q-24 24 -64 40.5t-73 16.5h-480q-33 0 -56.5 -23.5t-23.5 -56.5v-864q0 -33 23.5 -56.5t56.5 -23.5h736q33 0 56.5 23.5t23.5 56.5v608q0 33 -16.5 73t-40.5 64zM858 715q3 -2 5 -5l4 -6h-163v163l6 -4t5 -5zM896 16q0 -7 -4.5 -11.5t-11.5 -4.5h-736 q-7 0 -11.5 4.5t-4.5 11.5v864q0 7 4.5 11.5t11.5 4.5h480q4 0 8 -0.5t8 -1.5v-254h254q1 -4 1.5 -8t0.5 -8v-608z"
                transform="translate(0 930) scale(-1,1) rotate(180)"
              />
            </svg>
          </div>
        </div>
        <div
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "float": "left",
              "height": "40px",
              "margin": 0,
            }
          }
        >
          <div
            className="arduinosimulator_menu_item"
            onClick={[Function]}
            style={
              {
                "MozUserSelect": "none",
                "WebkitUserSelect": "none",
                "backgroundColor": "#F2F2F2",
                "border": "thin solid #F2F2F2",
                "cursor": "default",
                "display": "block",
                "fontFamily": "Arial",
                "fontSize": "15px",
                "height": "28px",
                "lineHeight": "32px",
                "marginLeft": "3px",
                "paddingBottom": "1px",
                "paddingTop": "3px",
                "userSelect": "none",
                "width": "28px",
              }
            }
          >
            <svg
              height={16}
              viewBox="0 0 576 512"
              width={16}
            >
              <path
                d="M572.694 292.093L500.27 416.248A63.997 63.997 0 0 1 444.989 448H45.025c-18.523 0-30.064-20.093-20.731-36.093l72.424-124.155A64 64 0 0 1 152 256h399.964c18.523 0 30.064 20.093 20.73 36.093zM152 224h328v-48c0-26.51-21.49-48-48-48H272l-64-64H48C21.49 64 0 85.49 0 112v278.046l69.077-118.418C86.214 242.25 117.989 224 152 224z"
                transform="translate(0 -60) scale(1,1.2)"
              />
            </svg>
          </div>
        </div>
        <div
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "float": "left",
              "height": "40px",
              "margin": 0,
            }
          }
        >
          <div
            className="arduinosimulator_menu_item"
            onClick={[Function]}
            style={
              {
                "MozUserSelect": "none",
                "WebkitUserSelect": "none",
                "backgroundColor": "#F2F2F2",
                "border": "thin solid #F2F2F2",
                "cursor": "default",
                "display": "block",
                "fontFamily": "Arial",
                "fontSize": "15px",
                "height": "28px",
                "lineHeight": "32px",
                "marginLeft": "3px",
                "paddingBottom": "1px",
                "paddingTop": "3px",
                "userSelect": "none",
                "width": "28px",
              }
            }
          >
            <svg
              height={16}
              viewBox="0 0 1000 1000"
              width={16}
            >
              <path
                d="M896 960h-896v-1024h1024v896zM512 832h128v-256h-128v256zM896 64h-768v768h64v-320h576v320h75l53 -53v-715z"
                transform="translate(-10 930) scale(-1,1) rotate(180)"
              />
            </svg>
          </div>
        </div>
        <div
          style={
            {
              "borderLeft": "thin solid #D3D3D3",
              "float": "left",
              "height": "100px",
              "marginLeft": "5px",
              "width": "1px",
            }
          }
        />
        <div
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "float": "left",
              "height": "40px",
              "margin": 0,
            }
          }
        >
          <div
            className="arduinosimulator_menu_item"
            onClick={[Function]}
            style={
              {
                "MozUserSelect": "none",
                "WebkitUserSelect": "none",
                "backgroundColor": "#F2F2F2",
                "border": "thin solid #F2F2F2",
                "cursor": "default",
                "display": "block",
                "fontFamily": "Arial",
                "fontSize": "15px",
                "height": "28px",
                "lineHeight": "32px",
                "marginLeft": "3px",
                "paddingBottom": "1px",
                "paddingTop": "3px",
                "userSelect": "none",
                "width": "28px",
              }
            }
          >
            <svg
              height={16}
              viewBox="0 0 1000 1000"
              width={16}
            >
              <path
                d="M762 -64q43 77 62 168t-9 168t-113.5 127.5t-253.5 46.5v-254l-384 384l384 384v-248q201 5 314.5 -73t148.5 -196t-4.5 -255.5t-144.5 -251.5z"
                transform="translate(0 930) scale(-1,1) rotate(180)"
              />
            </svg>
          </div>
        </div>
        <div
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "float": "left",
              "height": "40px",
              "margin": 0,
            }
          }
        >
          <div
            className="arduinosimulator_menu_item"
            onClick={[Function]}
            style={
              {
                "MozUserSelect": "none",
                "WebkitUserSelect": "none",
                "backgroundColor": "#F2F2F2",
                "border": "thin solid #F2F2F2",
                "cursor": "default",
                "display": "block",
                "fontFamily": "Arial",
                "fontSize": "15px",
                "height": "28px",
                "lineHeight": "32px",
                "marginLeft": "3px",
                "paddingBottom": "1px",
                "paddingTop": "3px",
                "userSelect": "none",
                "width": "28px",
              }
            }
          >
            <svg
              height={16}
              viewBox="0 0 1000 1000"
              width={16}
            >
              <path
                d="M576 712v248l384 -384l-384 -384v254q-168 4 -253.5 -46.5t-113.5 -127.5t-9 -168t62 -168q-105 114 -144.5 251.5t-4.5 255.5t148.5 196t314.5 73v0z"
                transform="translate(0 930) scale(-1,1) rotate(180)"
              />
            </svg>
          </div>
        </div>
        <div
          style={
            {
              "borderLeft": "thin solid #D3D3D3",
              "float": "left",
              "height": "100px",
              "marginLeft": "5px",
              "width": "1px",
            }
          }
        />
        <div
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "float": "left",
              "height": "40px",
              "margin": 0,
            }
          }
        >
          <div
            className="arduinosimulator_menu_item"
            onClick={[Function]}
            style={
              {
                "MozUserSelect": "none",
                "WebkitUserSelect": "none",
                "backgroundColor": "#F2F2F2",
                "border": "thin solid #F2F2F2",
                "cursor": "default",
                "display": "block",
                "fontFamily": "Arial",
                "fontSize": "15px",
                "height": "28px",
                "lineHeight": "32px",
                "marginLeft": "3px",
                "paddingBottom": "1px",
                "paddingTop": "3px",
                "userSelect": "none",
                "width": "28px",
              }
            }
          >
            <svg
              height={24}
              viewBox="0 0 24 24"
              width={24}
            >
              <path
                d="M11,5 C8.24078475,5 6,7.24078475 6,10 C6,12.7592153 8.24078475,15 11,15 C13.7592153,15 16,12.7592153 16,10 C16,7.24078475 13.7592153,5 11,5 Z M16.1517226,14.737509 L19.7071068,18.2928932 C20.0976311,18.6834175 20.0976311,19.3165825 19.7071068,19.7071068 C19.3165825,20.0976311 18.6834175,20.0976311 18.2928932,19.7071068 L14.5932864,16.0075 C13.5426554,16.6376337 12.3133603,17 11,17 C7.13621525,17 4,13.8637847 4,10 C4,6.13621525 7.13621525,3 11,3 C14.8637847,3 18,6.13621525 18,10 C18,11.826606 17.2990779,13.4906052 16.1517226,14.737509 Z"
                transform="translate(0 2)"
              />
            </svg>
          </div>
        </div>
        <div
          style={
            {
              "borderLeft": "thin solid #D3D3D3",
              "float": "left",
              "height": "100px",
              "marginLeft": "5px",
              "width": "1px",
            }
          }
        />
        <div
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "float": "left",
              "height": "40px",
              "margin": 0,
            }
          }
        >
          <div
            className="arduinosimulator_menu_item"
            onClick={[Function]}
            style={
              {
                "MozUserSelect": "none",
                "WebkitUserSelect": "none",
                "backgroundColor": "#F2F2F2",
                "border": "thin solid #F2F2F2",
                "cursor": "default",
                "display": "block",
                "fontFamily": "Arial",
                "fontSize": "15px",
                "height": "28px",
                "lineHeight": "32px",
                "marginLeft": "3px",
                "paddingBottom": "1px",
                "paddingTop": "3px",
                "userSelect": "none",
                "width": "28px",
              }
            }
          >
            <svg
              height={17}
              id="arduino_simulator_start_icon"
              viewBox="0 0 56.693 56.693"
              width={17}
            >
              <path
                d="M46.279,26.484L20.113,7.48c-0.957-0.695-2.228-0.796-3.278-0.258c-1.056,0.537-1.721,1.622-1.721,2.807  v38.009c0,1.184,0.665,2.268,1.721,2.805c0.45,0.23,0.939,0.344,1.429,0.344c0.652,0,1.303-0.203,1.85-0.602l26.166-19.006  c0.818-0.594,1.298-1.539,1.298-2.548C47.577,28.023,47.094,27.077,46.279,26.484z"
              />
            </svg>
          </div>
        </div>
        <div
          style={
            {
              "borderLeft": "thin solid #D3D3D3",
              "float": "left",
              "height": "100px",
              "marginLeft": "5px",
              "width": "1px",
            }
          }
        />
        <div
          style={
            {
              "alignItems": "center",
              "display": "flex",
              "float": "left",
              "height": "40px",
              "margin": 0,
            }
          }
        >
          <div
            style={
              {
                "MozUserSelect": "none",
                "WebkitUserSelect": "none",
                "cursor": "default",
                "float": "left",
                "fontFamily": "Arial",
                "fontSize": "13px",
                "lineHeight": 2.7,
                "marginLeft": "6px",
                "marginRight": "0px",
                "userSelect": "none",
              }
            }
          >
            Untitled.ino
          </div>
        </div>
      </div>
    </div>
  </div>,
  <input
    onChange={[Function]}
    style={
      {
        "display": "none",
      }
    }
    type="file"
  />,
]
`;
